drivers/adc/adc_ads131m02.c

```diff
@@ -0,0 +1,738 @@
+				   rx_buf, sizeof(rx_buf));
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (sys_get_be16(rx_buf) != ADS131M02_RESET_RSP) {
+		return -EIO;
+	}
+
+	k_msleep(ADS131M02_RESET_DELAY);
+
+	return 0;
+}
+
+#if defined CONFIG_PM_DEVICE
+static int ads131m02_pm(const struct device *dev, uint16_t cmd)
+{
+	int ret;
+	uint8_t tx_buf[3] = {0};
+	uint8_t rx_buf[3] = {0};
+
+	sys_put_be16(cmd, tx_buf);
+	ret = ads131m02_transceive(dev, tx_buf, sizeof(tx_buf),
+				   rx_buf, sizeof(rx_buf));
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (rx_buf[1] != cmd) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int ads131m02_pm_action(const struct device *dev,
+			       enum pm_device_action action)
+{
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		return ads131m02_pm(dev, ADS131M02_WAKEUP_CMD);
+	case PM_DEVICE_ACTION_SUSPEND:
+		return ads131m02_pm(dev, ADS131M02_STANDBY_CMD);
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+int ads131m02_set_adc_mode(const struct device *dev,
+			   enum ads131m02_adc_mode mode,
+			   enum ads131m02_gc_delay gc_delay)
+{
+	uint16_t temp = 0;
+
+	switch (mode) {
+	case ADS131M02_CONTINUOUS_MODE:
+		WRITE_BIT(temp, 8, 0);
+		break;
+	case ADS131M02_GLOBAL_CHOP_MODE:
+		temp |= ADS131M02_GC_MODE_MASK;
+		temp |= FIELD_PREP(ADS131M02_GC_DELAY_MASK, gc_delay);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ads131m02_reg_write(dev, ADS131M02_CFG_REG, temp);
+}
+
+int ads131m02_set_power_mode(const struct device *dev,
+			     enum ads131m02_adc_power_mode mode)
+{
+	int ret;
+	uint16_t temp = 0;
+	uint8_t buf[3] = {0};
+
+	ret = ads131m02_reg_read(dev, ADS131M02_CLOCK_REG, buf, sizeof(buf));
+	if (ret != 0) {
+		return ret;
+	}
+
+	temp = sys_get_be16(buf);
+	switch (mode) {
+	case ADS131M02_VLP:
+		WRITE_BIT(temp, 0, 0);
+		WRITE_BIT(temp, 1, 0);
+		break;
+	case ADS131M02_LP:
+		WRITE_BIT(temp, 0, 1);
+		WRITE_BIT(temp, 1, 0);
+		break;
+	case ADS131M02_HR:
+		WRITE_BIT(temp, 0, 1);
+		WRITE_BIT(temp, 1, 1);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ads131m02_reg_write(dev, ADS131M02_CLOCK_REG, temp);
+}
+
+static const struct adc_driver_api ads131m02_api = {
+	.channel_setup = ads131m02_channel_setup,
+	.read = ads131m02_read,
+	.ref_internal = ADS131M02_REF_INTERNAL,
+};
+
+static int ads131m02_init(const struct device *dev)
+{
+	int ret;
+	uint8_t buf[3] = {0};
+	const struct ads131m02_config *cfg = dev->config;
+	struct ads131m02_data *data = dev->data;
+
+	if (!ads131m02_check_device(&cfg->spi)) {
+		LOG_ERR("ADS131M02 is not ready");
+		return -ENODEV;
+	}
+
+	adc_context_init(&data->ctx);
+	k_sem_init(&data->acq_sem, 0, 1);
+	k_sem_init(&data->drdy_sem, 0, 1);
+
+	ret = ads131m02_configure_gpio(dev);
+	if (ret != 0) {
+		LOG_ERR("GPIO config failed %d", ret);
+		return ret;
+	}
+
+	ret = ads131m02_reg_read(dev, ADS131M02_ID_REG, buf, sizeof(buf));
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (buf[0] != ADS131M02_DEVICE_ID) {
+		LOG_ERR("Device ID mismatch");
```

Comment:
log the actual value, then you can distinguish between a missing or misassembled device or wrong part number at some extent

Reply:
done